pkg/operator/resources/apisplitter/api.go

```diff
@@ -0,0 +1,182 @@
+	"github.com/cortexlabs/cortex/pkg/lib/parallel"
+	"github.com/cortexlabs/cortex/pkg/operator/config"
+	"github.com/cortexlabs/cortex/pkg/operator/operator"
+	"github.com/cortexlabs/cortex/pkg/types/spec"
+	"github.com/cortexlabs/cortex/pkg/types/userconfig"
+	istioclientnetworking "istio.io/client-go/pkg/apis/networking/v1alpha3"
+)
+
+func UpdateAPI(apiConfig *userconfig.API, projectID string, force bool) (*spec.API, string, error) {
+	prevVirtualService, err := getK8sResources(apiConfig)
+	if err != nil {
+		return nil, "", err
+	}
+
+	api := spec.GetAPISpec(apiConfig, projectID, "")
+	if prevVirtualService == nil {
+		if err := config.AWS.UploadMsgpackToS3(api, config.Cluster.Bucket, api.Key); err != nil {
+			return nil, "", errors.Wrap(err, "upload api spec")
+		}
+		err = operator.AddAPIToAPIGateway(*api.Networking.Endpoint, api.Networking.APIGateway)
+		if err != nil {
+			go deleteK8sResources(api.Name)
+			return nil, "", err
+		}
+		if err := applyK8sResources(api, prevVirtualService); err != nil {
+			go deleteK8sResources(api.Name)
+			return nil, "", err
+		}
+
+		return api, fmt.Sprintf("creating %s", api.Name), nil
+	}
+
+	services, weight := getServicesWeightsTrafficSplitter(api)
+	if err != nil {
+		return nil, "", err
+	}
+
+	if !areVirtualServiceEqual(prevVirtualService, virtualServiceSpec(api, services, weight)) {
+		if err := config.AWS.UploadMsgpackToS3(api, config.Cluster.Bucket, api.Key); err != nil {
+			return nil, "", errors.Wrap(err, "upload api spec")
+		}
+		if err := applyK8sResources(api, prevVirtualService); err != nil {
+			return nil, "", err
+		}
+		if err := operator.UpdateAPIGatewayK8s(prevVirtualService, api); err != nil {
+			return nil, "", err
+		}
+		return api, fmt.Sprintf("updating %s", api.Name), nil
+	}
+	return api, fmt.Sprintf("%s is up to date", api.Name), nil
+}
+
+func DeleteAPI(apiName string, keepCache bool) error {
+	// best effort deletion, so don't handle error yet
+	virtualService, vsErr := config.K8s.GetVirtualService(operator.K8sName(apiName))
+	err := parallel.RunFirstErr(
+		func() error {
+			return vsErr
+		},
+		func() error {
+			return deleteK8sResources(apiName)
+		},
+		func() error {
+			if keepCache {
+				return nil
+			}
+			// best effort deletion
+			deleteS3Resources(apiName)
+			return nil
+		},
+		// delete API from API Gateway
+		func() error {
+			err := operator.RemoveAPIFromAPIGatewayK8s(virtualService)
+			if err != nil {
+				return err
+			}
+			return nil
+		},
+	)
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func getK8sResources(apiConfig *userconfig.API) (*istioclientnetworking.VirtualService, error) {
+	var virtualService *istioclientnetworking.VirtualService
+
+	virtualService, err := config.K8s.GetVirtualService(operator.K8sName(apiConfig.Name))
+	if err != nil {
+		return nil, err
+	}
+
+	return virtualService, err
+}
+
+func applyK8sResources(api *spec.API, prevVirtualService *istioclientnetworking.VirtualService) error {
```

Comment:
I thought it might be helpful to leave it because it makes `UpdateAPI` "consistent" across different Kinds.

Reply:
Yea I think it makes sense in theory. At this point, since we don't have an interface to enforce consistency across the different kinds, I don't think it is important to create functions to maintain consistency.